src/redis-cli.c

```diff
@@ -1909,6 +1909,8 @@ static void usage(int err) {
 "Examples:\n"
 "  cat /etc/passwd | redis-cli -x set mypasswd\n"
 "  redis-cli get mypasswd\n"
+"  cat /etc/passwd | redis-cli -X passwd_tag set mypasswd2 passwd_tag\n"
+"  redis-cli get mypasswd2\n"
```

Comment:
i don't think we need that second line (let's keep it short).
it's enough to show the redis-cli syntax, and the user can figure out on his own what it does.
maybe add one to show the usage of DUMP piped into RESTORE using `-x`?

Reply:
I found out that there is always a `ERR DUMP payload version` in common usage.
should i mention it in the help text? (or maybe i did something wrong...?)

```
[root@binblog redis]# src/redis-cli sadd set 1 2 3 a b c
(integer) 6
[root@binblog redis]# src/redis-cli dump set > set.dump

# -x option
[root@binblog redis]# src/redis-cli -x restore set2 0 < set.dump 
(error) ERR DUMP payload version or checksum are wrong

[root@binblog redis]# head -c -1 set.dump | src/redis-cli -x restore set3 0
OK
[root@binblog redis]# src/redis-cli scard set3
(integer) 6

---------------------
# -X option
[root@binblog redis]# src/redis-cli -X dump_tag restore set4 0 dump_tag < set.dump 
(error) ERR DUMP payload version or checksum are wrong

[root@binblog redis]# head -c -1 set.dump | src/redis-cli -X dump_tag restore set3 0 dump_tag
OK

[root@binblog redis]# head -c -1 set.dump | src/redis-cli -X dump_tag restore set3 0 dump_tag replace
OK
```

I may change the example to: 
```
"Examples:\n"
"  cat /etc/passwd | redis-cli -x set mypasswd\n"
"  redis-cli dump key > key.dump && redis-cli -x restore key2 0 < key.dump\n"
"  cat /etc/passwd | redis-cli -X passwd_tag set mypasswd passwd_tag\n"
"  redis-cli dump key > key.dump && redis-cli -X dump_tag restore key2 0 dump_tag replace < key.dump\n"
```

Reply:
@enjoy-binbin is that because DUMP adds a newline at the end of the payload and that's written to the file?
seems like a bug, can you please look into it? maybe `--raw` will help? 

your example seems right, but maybe excessive, i don't think we need to provide two examples for both `-x` and `-X`.
IMHO it is enough for instance to use the password example for `-x` and the dump example for `-X`

Reply:
DUMP is ok, the newline was added by `> set.dump`

Reply:
are you sure? that doesn't sound right.
so if i `cat file.jpg > newfile.jpg` it adds a newline? doesn't sound right.. please look closer at redis-cli

Reply:
I tried to comment out this line break in redis-cli
```c
    case REDIS_REPLY_STRING:
    case REDIS_REPLY_VERB:
        /* If you are producing output for the standard output we want
        * a more interesting output with quoted characters and so forth,
        * unless it's a verbatim string type. */
        if (r->type == REDIS_REPLY_STRING) {
            out = sdscatrepr(out,r->str,r->len);
            // out = sdscat(out,"\n");  <------------- this line
        } else {
            out = sdscatlen(out,r->str,r->len);
            out = sdscat(out,"\n");
        }
```

Maybe the output is not very friendly :)
```
# and indeed, the newline was removed
[root@binblog redis]# src/redis-cli dump set
"\x02\x06\x01b\x01a\xc0\x01\xc0\x02\xc0\x03\x01c\n\x00\xc4\xf2\xae\xc8\xe8\x03\xd7\xf1"[root@binblog redis]#

# but it still output with a newline...
[root@binblog redis]# src/redis-cli dump set > set.dump
[root@binblog redis]# cat set.dump
ba@󿿣
®
[root@binblog redis]# head set.dump
ba@󿿣
®
[root@binblog redis]# src/redis-cli -x restore new_set2 0 < set.dump
(error) ERR DUMP payload version or checksum are wrong

# we remove the newline
[root@binblog redis]# head -c -1 set.dump
ba@󿿣
®[root@binblog redis]#   --------- focus on the left ®
[root@binblog redis]# head -c -1 set.dump > set.dump2
[root@binblog redis]# cat set.dump2 
ba@󿿣
®[root@binblog redis]#   --------- focus on the left ®
[root@binblog redis]# src/redis-cli -x restore new_set3 0 < set.dump2                                                                                                         
OK

# string reply
[root@binblog redis]# src/redis-cli get a
"a"[root@binblog redis]# src/redis-cli get a > a.txt
[root@binblog redis]# cat a.txt 
a
[root@binblog redis]#
```

Reply:
if i use --raw
```
    if (verbatim) {
        out = cliFormatReplyRaw(reply);
    } else if (mode == OUTPUT_STANDARD) {
        out = cliFormatReplyTTY(reply, "");
    } else if (mode == OUTPUT_RAW) {
        out = cliFormatReplyRaw(reply);
        // out = sdscatsds(out, config.cmd_delim);   --------- comment this line
```

and try, it work
```
[root@binblog redis]# src/redis-cli --raw dump set > set.dump
[root@binblog redis]# cat set.dump
ba@󿿣
®[root@binblog redis]# src/redis-cli -x restore new_set4 0 < set.dump
OK
```

Reply:
ok, so it's redis-cli, not the system's `>`, and we already have a way around it.
so yes, let's just include `-D ""` in the example

Reply:
i want to write some tests like this.. but failed.. always `ERR DUMP`
```
    test "DUMP RESTORE with -x option" {
        run_cli lpush mylist a b c d e
        set res [run_cli llen mylist]
        # r config set sanitize-dump-payload no

        set dump [run_cli -D "" --raw dump mylist]
        set dump2 [run_cli dump mylist]

        set tmpfile [write_tmpfile $dump]
        # set tmpfile2 [write_tmpfile $dump2]

        set res [run_cli_with_input_file x $tmpfile restore mylist2 0]
        puts "try 1: $res"

        set res [run_cli_with_input_pipe x "echo -n $dump" restore mylist2 0]
        puts "try 2: $res"

        set res [run_cli_with_input_pipe X "echo -e $dump" restore mylist2 0 tag replace]
        puts "try 3: $res"

        set res [run_cli llen mylist2]
        puts "llen: $res"

        file delete $tmpfile
    }
```

Reply:
don't know..
but maybe it's a better idea to either rely on the shell (do `exec sh .... | ....`), or use the piping features of [exec](https://www.tcl.tk/man/tcl8.5/tutorial/Tcl26.html) to pipe these into each other without using files?